src/Compilers/CSharp/Portable/Symbols/AliasSymbol.cs

```diff
@@ -259,7 +207,132 @@ public override TResult Accept<TResult>(CSharpSymbolVisitor<TResult> visitor)
         }
 
         // basesBeingResolved is only used to break circular references.
-        internal NamespaceOrTypeSymbol GetAliasTarget(ConsList<TypeSymbol>? basesBeingResolved)
+        internal abstract NamespaceOrTypeSymbol GetAliasTarget(ConsList<TypeSymbol>? basesBeingResolved);
+
+        internal void CheckConstraints(BindingDiagnosticBag diagnostics)
+        {
+            var target = this.Target as TypeSymbol;
+            if ((object?)target != null && Locations.Length > 0)
+            {
+                var corLibrary = this.ContainingAssembly.CorLibrary;
+                var conversions = new TypeConversions(corLibrary);
+                target.CheckAllConstraints(DeclaringCompilation, conversions, Locations[0], diagnostics);
+            }
+        }
+
+        public override bool Equals(Symbol? obj, TypeCompareKind compareKind)
+        {
+            if (ReferenceEquals(this, obj))
+            {
+                return true;
+            }
+
+            if (ReferenceEquals(obj, null))
+            {
+                return false;
+            }
+
+            AliasSymbol? other = obj as AliasSymbol;
+
+            return (object?)other != null &&
+                Equals(this.Locations.FirstOrDefault(), other.Locations.FirstOrDefault()) &&
```

Comment:
This is a pre-existing issue (or be by design), but I'm not certain how this equality algorithm is correct. It would return true for multiple locationless aliases of different names that come from same assembly. For example, it appears that it would return true for `alias.ToNewSubmission(comp).Equals(alias)`. Is this intended?

Reply:
It also appears to be incorrect with respect to `GetHashCode`, as that will use the name if the location is empty, and equality does not take the name into consideration.

---
In reply to: [583953172](https://github.com/dotnet/roslyn/pull/51505#discussion_r583953172) [](ancestors = 583953172)

Reply:
There is no change in implementation of these both functions. I didn't even change their location. It is a preference of the diff tool to show this as a move. 
As to the question whether the implementation is correct. I believe the only time location is missing is when the alias represents a fake ```global``` alias. In all other use cases location defines the name. GetHashCode uses name probably just an alternative to a constant value. If you think there is a usage example (not a hypothetical thing that someone theoretically could do) where that is not true, please open an issue.
 

---
In reply to: [583955291](https://github.com/dotnet/roslyn/pull/51505#discussion_r583955291) [](ancestors = 583955291,583953172)

Reply:
My concern is that, because `Equals` and `GetHashCode` don't follow the rules for how they should behave, we could see some odd behavior. Perhaps an assert in the constructor that the only time `_locations` is empty is when the name is `global` to ensure that we have correct behavior?

---
In reply to: [583964176](https://github.com/dotnet/roslyn/pull/51505#discussion_r583964176) [](ancestors = 583964176,583955291,583953172)

Reply:
> Perhaps an assert in the constructor that the only time _locations is empty is when the name is global to ensure that we have correct behavior?

This is outside of the scope of this PR. I believe it doesn't introduce any observable behavior change.

---
In reply to: [583966516](https://github.com/dotnet/roslyn/pull/51505#discussion_r583966516) [](ancestors = 583966516,583964176,583955291,583953172)